zhmcclient/_adapter.py

```diff
@@ -0,0 +1,178 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+An **Adapter** object represents a single adapter for
+a Dynamic Partition Manager-enabled (DPM) system.
+Adapters on a z Systems or LinuxONE system fall into four categories:
+Network, Storage, Accelerators, and Cryptos.
+Each adapter type plays a specific role in communication, or data transfer,
+for system partitions and the applications that run in them.
+Objects of this class are not provided when the CPC is not in DPM mode.
+
+An adapter is always contained in a CPC.
+"""
+
+from __future__ import absolute_import
+
+from ._manager import BaseManager
+from ._resource import BaseResource
+
+__all__ = ['AdapterManager', 'Adapter']
+
+
+class AdapterManager(BaseManager):
+    """
+    Manager object for Adapters. This manager object is scoped to the
+    adapters of a particular CPC.
+
+    Derived from :class:`~zhmcclient.BaseManager`; see there for common methods
+    and attributes.
+    """
+
+    def __init__(self, cpc):
+        """
+        Parameters:
+
+          cpc (:class:`~zhmcclient.Cpc`):
+            CPC defining the scope for this manager object.
+        """
+        super(AdapterManager, self).__init__(cpc)
+
+    @property
+    def cpc(self):
+        """
+        :class:`~zhmcclient.Cpc`: Parent object (CPC) defining the scope for
+        this manager object.
+        """
+        return self._parent
+
+    def list(self, full_properties=False):
+        """
+        List the adapters in scope of this manager object.
+
+        Parameters:
+
+          full_properties (bool):
+            Controls whether the full set of resource properties should be
+            retrieved, vs. only the short set as returned by the list
+            operation.
+
+        Returns:
+
+          : A list of :class:`~zhmcclient.Adapter` objects.
+
+        Raises:
+
+          :exc:`~zhmcclient.HTTPError`
+          :exc:`~zhmcclient.ParseError`
+          :exc:`~zhmcclient.AuthError`
+          :exc:`~zhmcclient.ConnectionError`
+        """
+        cpc_uri = self.cpc.get_property('object-uri')
+        adapters_res = self.session.get(cpc_uri + '/adapters')
+        adapter_list = []
+        if adapters_res:
+            adapter_items = adapters_res['adapters']
+            for adapter_props in adapter_items:
+                adapter = Adapter(self, adapter_props['object-uri'],
+                                  adapter_props)
+                if full_properties:
+                    adapter.pull_full_properties()
+                adapter_list.append(adapter)
+        return adapter_list
+
+    def create_hipersocket(self, properties):
+        """
+        Create and configures a HiperSockets adapter
+        with the specified resource properties.
+
+        Parameters:
+
+          properties (dict): Properties for the new adapter.
+
+        Returns:
+
+          string: The resource URI of the new adapter.
```

Comment:
The newly created HiperSockets adapter should be useable like if it was returned by the `list()` method. This means we should return an `Adapter` object that has "object-uri" set as its only property. This allows users to use its methods, e.g. `get_property()` which would fill the other properties as needed.